8 задание

CyberJucifer opened the pull request:
Вывод результата теста

CyberJucifer committed:
8 задание

Вывод результата теста

VidgarVii commented on app/helpers/test_passages_helper.rb:
```diff
@@ -0,0 +1,9 @@
+  private
+
+  def result_in_percents(test_passage)
+    (test_passage.correct_questions.to_f/test_passage.test.questions.count.to_f * 100.0).to_i
```
все числа преобразовывать во float не обязательно. Достоточно одного такого числа. И руби сам результат выражение сделает с плав. точкой

VidgarVii commented on app/helpers/test_passages_helper.rb:
```diff
@@ -0,0 +1,9 @@
+module TestPassagesHelper
+
+  private
```
private не нужен

VidgarVii commented on app/models/test.rb:
```diff
@@ -2,9 +2,9 @@ class Test < ApplicationRecord
   belongs_to :category, optional: true
   belongs_to :creator, class_name: 'User', foreign_key: :creator_id
 
+  has_many :test_passages
```
следует добавить dependent

VidgarVii commented on app/models/test_passage.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  def correct_answer?(answer_ids)
+    correct_answers.ids.sort == answer_ids.map(&:to_i).sort
```
если пользователь не выберет ответ то будет ошибка

VidgarVii commented on app/models/user.rb:
```diff
@@ -1,6 +1,6 @@
-  has_many :user_tests, dependent: :destroy
-  has_many :tests, through: :user_tests
+  has_many :test_passages
+  has_many :tests, through: :test_passages
```
нужен dependent
иначе при удалении юзера останутся не консистентные данные в test_passages

VidgarVii commented on app/models/user.rb:
```diff
@@ -9,4 +9,8 @@ class User < ApplicationRecord
   end
+
+  def test_passage(test)
+    test_passages.order(id: :desc).find_by(test_id: test.id)
```
find_by(test: test)

ООП это про декларативность. Мы говорим что делать, а не как.

CyberJucifer committed:
Изменил БД, исправил миграции

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -0,0 +1,20 @@
+class TestPassagesController < ApplicationController
+
+  before_action :find_test_passage, only: %i[show update result]
```
https://github.com/rubocop-hq/rails-style-guide#lexically-scoped-action-filter

VidgarVii commented on app/views/test_passages/result.html.erb:
```diff
@@ -0,0 +1,9 @@
+  <p>Success! Your result: <span style="color:green"><%= result %>%</span></p>
+<% else %>
+  <p>Failure! Your result: <span style="color:red"><%= result %>%</span></p>
+<% end %>
```
result >= 85 это бизнес логика. Ей не место в представлении
Пройден ли был тест, надо спрашивать у объекта @test_passage

VidgarVii commented on db/migrate/20210114155759_add_creator_and_category_to_tests.rb:
```diff
@@ -0,0 +1,6 @@
+class AddCreatorAndCategoryToTests < ActiveRecord::Migration[5.2]
+  def change
+    add_reference(:tests, :creator, foreign_key: { to_table: :tests } )
+    add_reference(:tests, :category, foreign_key: { to_table: :categories } )
```
тут to_table нет смысла объявлять. Так как у тебя название таблицы categories идентично названию FK

VidgarVii commented on db/migrate/20210114155759_add_creator_and_category_to_tests.rb:
```diff
@@ -0,0 +1,6 @@
+class AddCreatorAndCategoryToTests < ActiveRecord::Migration[5.2]
+  def change
+    add_reference(:tests, :creator, foreign_key: { to_table: :tests } )
```
тут у тебя связь сам к себе. Создатель теста тест?

CyberJucifer committed:
Удалил папку .idea из репозитория

CyberJucifer committed:
Поправил представление, миграцию.

Добавил метод проверки успеха прохождения теста в модель test_passage.rb. Поправил миграцию добавления создателя/категории к тесту.
